mesonbuild/environment.py

```diff
@@ -563,13 +585,29 @@ def __init__(self, source_dir, build_dir, options):
             if 'target_machine' in config:
                 machines.target = MachineInfo.from_literal(config['target_machine'])
             paths.host = Directories(**config.get('paths', {}))
+            load_options('project options', user_options)
+            meson_options.host = {}
+            load_options('built-in options', meson_options.host)
 
         ## "freeze" now initialized configuration, and "save" to the class.
 
         self.machines = machines.default_missing()
         self.binaries = binaries.default_missing()
         self.properties = properties.default_missing()
         self.paths = paths.default_missing()
+        self.user_options = user_options
+        self.meson_options = meson_options.default_missing()
+
+        # Ensure that no paths are passed via built-in options:
+        if '' in self.meson_options.host:
+            for each in ['bindir', 'datadir', 'includedir', 'infodir', 'libdir',
+                         'libexecdir', 'localedir', 'localstatedir', 'mandir',
+                         'prefix', 'sbindir', 'sharedstatedir', 'sysconfdir']:
```

Comment:
mconf.py already has a dir_option_names list, maybe we should move that list into coredata to maintain it in a single place?